direct/tpm2/policy_test.go

```diff
@@ -281,6 +282,189 @@ func TestPolicyOrUpdate(t *testing.T) {
+		}
+	}()
+
+	PCRs, err := CreatePCRSelection([]int{0, 1, 2, 3, 7})
+	if err != nil {
+		t.Fatalf("Failed to create PCRSelection")
+	}
+	selection := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      tpm.AlgSHA1,
+				PCRSelect: PCRs,
+			},
+		},
+	}
+
+	pcrRead := PCRRead{
+		PCRSelectionIn: selection,
+	}
+
+	pcrReadRsp, err := pcrRead.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Failed to read PCRSelection")
+	}
+
+	var expectedVal []byte
+	for _, digest := range pcrReadRsp.PCRValues.Digests {
+		expectedVal = append(expectedVal, digest.Buffer...)
+	}
+	t.Logf("expectedVal=%x", expectedVal)
+
+	// Hash algorithm must match the one in PolicySession.
+	expectedDigest := sha1.Sum(expectedVal)
+	t.Logf("expectedDigest=%x", expectedDigest)
+
+	wrongCase := append(expectedVal, []byte("wrongCase")...)
+	wrongDigest := sha1.Sum(wrongCase)
+
+	tests := []struct {
+		desc           string
+		expectedDigest []byte
+		expectedResult bool
+	}{
+		{desc: "success: expectedDigest passed in expectedDigest", expectedDigest: expectedDigest[:], expectedResult: true},
+		{desc: "fail: wrongDigest passed in expectedDigest", expectedDigest: wrongDigest[:], expectedResult: false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			policyPCR := PolicyPCR{
+				PolicySession: sess.Handle(),
+				PcrDigest: tpm2b.Digest{
+					Buffer: tt.expectedDigest,
+				},
+				Pcrs: selection,
+			}
+
+			if err := policyPCR.Execute(thetpm); err != nil {
+				t.Fatalf("executing PolicyPCR: %v", err)
+			}
+
+			pgd := PolicyGetDigest{
+				PolicySession: sess.Handle(),
+			}
+			want, err := pgd.Execute(thetpm)
+			if err != nil {
+				t.Fatalf("executing PolicyGetDigest: %v", err)
+			}
+
+			// Use the policy helper to calculate the same policy
+			pol, err := NewPolicyCalculator(tpm.AlgSHA1)
+			if err != nil {
+				t.Fatalf("creating policy calculator: %v", err)
+			}
+			policyPCR.Update(pol)
+			got := pol.Hash()
+
+			if bytes.Equal(got.Digest, want.PolicyDigest.Buffer) != tt.expectedResult {
+				t.Errorf("policyPCR.Hash() = %x,\nwant %x", got.Digest, want.PolicyDigest.Buffer)
+			}
+		})
+	}
+}
+
+// Testing void input for PcrDigest, from TPM 2.0 Part 3, 23.7.
+func TestPolicyPCRVoidPcrDigest(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	// Use a trial session to calculate this policy
+	sess, cleanup2, err := PolicySession(thetpm, tpm.AlgSHA1, 16, Trial())
+	if err != nil {
+		t.Fatalf("setting up policy session: %v", err)
+	}
+	defer func() {
+		t.Helper()
+		if err := cleanup2(); err != nil {
+			t.Errorf("cleaning up policy session: %v", err)
+		}
+	}()
+
+	PCRs, err := CreatePCRSelection([]int{0, 3, 7})
+	if err != nil {
+		t.Fatalf("Failed to create PCRSelection")
+	}
+	selection := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      tpm.AlgSHA1,
```

Comment:
Resolved by having the void testcase test SHA256.

Reply:
One thing could be to add a nested loop that goes through the algorithms we support, so that we get coverage on hash calculations as well. 
something like:
```
        tpmAlgs := []tpm.HashAlg{tpm.AlgSHA1, tpm.AlgSHA256}
	tests := []struct {
		desc           string
		expectedDigest []byte
		expectedResult bool
	}{
		{desc: "success: expectedDigest passed in expectedDigest", expectedDigest: expectedDigest[:], expectedResult: true},
		{desc: "fail: wrongDigest passed in expectedDigest", expectedDigest: wrongDigest[:], expectedResult: false},
	}
	for _, tt := range tests {	tests := []struct {
		desc           string
		expectedDigest []byte
		expectedResult bool
	}{
		{desc: "success: expectedDigest passed in expectedDigest", expectedDigest: expectedDigest[:], expectedResult: true},
		{desc: "fail: wrongDigest passed in expectedDigest", expectedDigest: wrongDigest[:], expectedResult: false},
	}
	for _, tt := range tests {
          for _, tpmAlg := range tpmAlgs {
// test logic here.
```

Reply:
The reason I did not suggest this is that since the purpose of this is to test the software stack and not the TPM, we don't have to be exhaustive about covering every single algorithm in every single case. We just need to make sure that go-tpm marshals commands and unmarshals responses correctly, and IMHO as long as 2 or more hash algorithms are represented across all the tests for a command, we have plenty of coverage of that. That said, if you disagree, I won't spend a lot of effort arguing against even more thorough testing :)

Reply:
I do not have strong feelings on the matter except that we should calculate a SHA256 digest somewhere with the PolicyCalculator. If we already have, great! Otherwise in these tests, I believe it's all SHA1 now.